### ISSUE TITLE
Error when processing OSM-extracted data frames

### ISSUE THREAD

**Issue by ekatef** [pz-max/pypsa-africa#278]
Downloading and preprocessing of OSM data throws an error for some regions which seems to indicate some discrepancies between the actual and expected structure of an extracted data frames. The error for download may be temporary resolved by increasing a requested area (e.g. adding some countries to an initially requested one) but reappears after data cleaning is run on loaded data.

The problem was reported by @EmreYorat for Armenia and reproduced for some countries of Central Asia and Central America. 

## Minimal reproducible example

Request of Armenia data by defining `countries: ["AM"]` in config.yaml and running `snakemake -j 1 download_osm_data` leads to `KeyError: 'lonlat'` triggered by a call `df_all_feature["lonlat"].apply` in download_osm_data.py:

https://github.com/pypsa-meets-africa/pypsa-africa/blob/56ac15f7edb1b1c6183f88cc85d7f32f1af66402/scripts/download_osm_data.py#L445-L450

A workaround is to set `countries: ["AZ", "AM", "GE"]` which allows to download OSM data successfully. However, running a cleaning script `snakemake -j 1 clean_osm_data` on the loaded data leads to another error: `ValueError: invalid literal for int() with base 10: '1/3'` which also is connected with subsetting of the OSM-extracted data frame ` df_all_lines["circuits"] = df_all_lines["circuits"]`

https://github.com/pypsa-meets-africa/pypsa-africa/blob/56ac15f7edb1b1c6183f88cc85d7f32f1af66402/scripts/clean_osm_data.py#L332-L335

## Error Messages

The whole error messages chain looks as follows:

### During data load
`download_osm_data` with `countries: ["AM"]`

```bash
Traceback (most recent call last):
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 3361, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 76, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 108, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/hashtable_class_helper.pxi", line 5198, in pandas._libs.hashtable.PyObjectHashTable.get_item
  File "pandas/_libs/hashtable_class_helper.pxi", line 5206, in pandas._libs.hashtable.PyObjectHashTable.get_item
KeyError: 'lonlat'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "~/Documents/_github_/pypsa-africa/.snakemake/scripts/tmpdx2v6wpw.download_osm_data.py", line 728, in <module>
    process_data(
  File "~/Documents/_github_/pypsa-africa/.snakemake/scripts/tmpdx2v6wpw.download_osm_data.py", line 595, in process_data
    output_csv_geojson(
  File "~/Documents/_github_/pypsa-africa/.snakemake/scripts/tmpdx2v6wpw.download_osm_data.py", line 452, in output_csv_geojson
    is_linestring = df_all_feature["lonlat"].apply(
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/frame.py", line 3458, in __getitem__
    indexer = self.columns.get_loc(key)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 3363, in get_loc
    raise KeyError(key) from err
KeyError: 'lonlat'
```

### During data cleaning
`clean_osm_data` with `countries: ["AZ", "AM", "GE"]`

```bash
Traceback (most recent call last):
  File "repos/pypsa-africa/.snakemake/scripts/tmpfkw0af9k.clean_osm_data.py", line 593, in <module>
    clean_data(
  File "repos/pypsa-africa/.snakemake/scripts/tmpfkw0af9k.clean_osm_data.py", line 468, in clean_data
    df_all_lines = integrate_lines_df(df_all_lines)
  File "repos/pypsa-africa/.snakemake/scripts/tmpfkw0af9k.clean_osm_data.py", line 339, in integrate_lines_df
    df_all_lines["circuits"] = df_all_lines["circuits"].astype(int)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/generic.py", line 5815, in astype
    new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/internals/managers.py", line 418, in astype
    return self.apply("astype", dtype=dtype, copy=copy, errors=errors)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/internals/managers.py", line 327, in apply
    applied = getattr(b, f)(**kwargs)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/internals/blocks.py", line 591, in astype
    new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/dtypes/cast.py", line 1309, in astype_array_safe
    new_values = astype_array(values, dtype, copy=copy)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/dtypes/cast.py", line 1257, in astype_array
    values = astype_nansafe(values, dtype, copy=copy)
  File "~/miniconda3/envs/pypsa-africa/lib/python3.9/site-packages/pandas/core/dtypes/cast.py", line 1174, in astype_nansafe
    return lib.astype_intsafe(arr, dtype)
  File "pandas/_libs/lib.pyx", line 679, in pandas._libs.lib.astype_intsafe
ValueError: invalid literal for int() with base 10: '1/3'
```

## Checklist

- [x] I am using the branch which is up-to-date with the current [`main`](https://github.com/pypsa-meets-africa/pypsa-africa/tree/main) branch or the latest [release](https://github.com/pypsa-meets-africa/pypsa-africa/releases).
- [x] I am running on an up-to-date [`pypsa-africa` environment](https://github.com/pypsa-meets-africa/pypsa-africa/blob/main/envs/environment.yaml).

**Comment by ekatef** [comment]
## Update

### Data load error

An error during OSM data download seems to be caused by an empty cables file. When it's a case the `df_all_feature` typically contains only `"Type"` and `"Country"` columns and an attempt to select a `"lonlat"` column results in an error.

### Data cleaning error

An error during OSM data cleaning is connected with a special notation that is sometimes used as a value of the OSM [circuits](https://wiki.openstreetmap.org/wiki/Key:circuits) tag related to the [line](https://wiki.openstreetmap.org/wiki/Tag%3Apower%3Dline) feature. According to the OSM rules the circuits field should be defined keeping in mind:

> The number of separate electrical circuits carried by this power line. For the common 3-phase systems, it's no. of conductors/3

So, usually the circuits number is defined as an integer or a combination of the integers. But there are [also cases](https://taginfo.openstreetmap.org/keys/?key=circuits#values) when it's set as `"1/3"` which causes an error when trying to transform the value into an integer.  In particular, that is the case for Georgia, Kazakhstan, Russia and some European countries.

**Comment by davide-f** [comment]
Awesome!

## Data load error

So the lonlat problem should be easy to fix by specifying the structure of the dataframe.

## Data cleaning error
Regarding the cleaning phase, it's a bit trickier: Are there really non-three phase high voltage lines or is it an error? This needs a bit of validation also with data from local transmission operators.
May you collect examples in which the number of conductors is not a multipe of three?
In such cases, may you verify whether the corresponding value of the circuits column is provided?

In any case, we shall add a logging warning listing the countries in which the unexpected number of conductors occurs

**Comment by ekatef** [comment]
Hey @davide-f! Sorry for taking a pause.

## Data loading error

> So the lonlat problem should be easy to fix by specifying the structure of the dataframe.

Great, then I'll propose a fix which check structure of the data frame. Thank you for the idea!

## Data cleaning error

> Are there really non-three phase high voltage lines or is it an error? This needs a bit of validation also with data from local transmission operators.
> May you collect examples in which the number of conductors is not a multipe of three?
> In such cases, may you verify whether the corresponding value of the circuits column is provided?

Right at the moment I don't have power grid data detailed enough to look into the "1/3-circuits" issue. But pure exploration of OSM data seems to give a rather consistent picture: if number of cables is **1/3** then the number of circuits is most often **1** and visualisation of such transmission lines gives a typical pattern like this:

<img width="184" alt="image" src="https://user-images.githubusercontent.com/30229437/157469758-d7ac609b-0673-40c6-9934-695e289f08a1.png">

Here "1/3-circuits" are in red while all other lines are blue. I suppose these red patterns are some local components of the power grid which are real but probably not particularly relevant for the model. What do you think?

**Comment by davide-f** [comment]
Mmmm very interesting!
So, I have investigated the issue using also https://overpass-turbo.eu/ that enables performing some cool comparisons.

That problems occurs for:
- HVDC lines: in HVDC lines, there is no three phase, obviously, thus the filtering shall account for AC or DC lines
![image](https://user-images.githubusercontent.com/67809479/157549456-5ff2c9cc-391d-48ed-9147-19ba9f8312d9.png)
- actual single phase AC lines. There are some AC lines that are single-phase; often they are denoted as are three single phase lines instead of a three phase line, though this is more the case for europe. However they are scarce and usually very short; if we use a large-enough tolerance for gathering the substation, we neglect the issue. The easy way for the AC lines with non-multiple-of-three cables is to disregard them, though, it would be nice to account for them in the modelling, but it is not easy.

In short:
- Convert cables into circuits only where circuits is not available
- For HVDC lines the cables/3 shall not be done but circuits should be used instead and if not available use cables/2
- For HVAC lines that have number of cables non multiple of three, for the moment, let's remove them, but put a warning regarding the maximum length of the line that has been removed and how many lines have been removed. We should track this challange as when we go global this challenge may be larger.

Next steps, would be to investigate the appropriate inclusion of single phase HVAC lines

**Comment by pz-max** [comment]
> * For HVAC lines that have number of cables non multiple of three, for the moment, let's remove them, but put a warning regarding the maximum length of the line that has been removed and how many lines have been removed. We should track this challange as when we go global this challenge may be larger.

Openstreetmap data is tagged by people. Some HVAC lines have beyond the electricity carrying power lines also lightning protection rods. See and read more about it here:  https://www.mto.com.tr/single-post/high-voltage-power-transmission-lines-lightning-protection. I think this is a major cause for such non-multiple of three datasets.
![image](https://user-images.githubusercontent.com/61968949/157559789-4b2184db-167a-43d0-a3b1-f28010877eb9.png)

**Comment by ekatef** [comment]
Hey @davide-f , hey @pz-max!

@davide-f Thank you so much for analysis! I'll be happy to make a proper check of the global OSM data and the guidance you have created will be very helpful for that.

@pz-max Many thanks for the comment. It makes a big picture much clearer.  Actually, lighting protection cables are being taking into account while cleaning OSM data

https://github.com/pypsa-meets-africa/pypsa-africa/blob/c994217e2c5fff891e39dde5fe6da6d27757c498/scripts/clean_osm_data.py#L324-L330

Absolutely agree that it's definitely worth to remember that lighting protection exists when proposing any further changes. 

Generally, it seems that a problem is a little deeper than a simple type conversion error. I think it makes sense to make first a quick fix of the errors appearing when loading and cleaning OSM data. (I'll try to propose such a fix in #287) And then to do a deeper analysis to verify and validate HVDC lines as the issue #286 proposes.

**Comment by davide-f** [comment]
Let's go for it! I agree, moreover, the first step would surely help clarifying the issue for the subsequent step.
Anyway, I think the issue #286 is very linked to the #287. You may see it along the way, it may be efficient to tackle them together, yet do as you please :)

Thanks!
